File: 프로그래머스/2/42587. 프로세스/프로세스.js
function solution(priorities, location) {
    const queue = priorities.map((p, idx) => [p, idx]);
    let count = 0;

    while (queue.length > 0) {
        const [currentPriority, currentIndex] = queue.shift();

        const hasHigher = queue.some(([p, _]) => p > currentPriority);
        if (hasHigher) {
            queue.push([currentPriority, currentIndex]);
        } else {
            count++;
            if (currentIndex === location) {
                return count;
            }
        }
    }
}

File: 프로그래머스/2/42587. 프로세스/프로세스.java
import java.util.Deque;
import java.util.ArrayDeque;

class Process {
    int priority;
    int index;
    
    Process(int priority, int index){
        this.priority = priority;
        this.index = index;
    }
}

class Solution {
    public int solution(int[] priorities, int location) {
        Deque<Process> queue = new ArrayDeque();
        int n = priorities.length;
        for (int i = 0; i < n; i ++){
            queue.offerLast(new Process(priorities[i], i));
        }
        
        int count = 0;
        
        while (!queue.isEmpty()){
            Process current = queue.pollFirst();
            boolean hasHigher = false;
            
            for (Process p: queue){
                if (p.priority > current.priority){
                    hasHigher = true;
                    break;
                }
            }
            
            if (hasHigher){
                queue.offerLast(current);
            }else{
                count ++;
                if (current.index == location){
                    return count;
                }
            }
        }
        
        return count;
    }
}